Chef Infra Client's Compliance Phase docs. (#2781)
chef/chef-web-docs: content/chef_compliance_phase.md

## Overview

### Component Architecture
```
 ┌──────────────────────┐    ┌──────────────────────┐    ┌─────────────────────┐
 │  Chef Infra Client   │    │  Chef Infra Server   │    │    Chef Automate    │
 │                      │    │      (optional)      │    │                     │
 │ ┌──────────────────┐ │    │                      │    │                     │
 │ │                  │◀┼────┼──────────────────────┼────│  Profiles           │
 │ │ Compliance Phase │ │    │                      │    │                     │
 │ │                  │─┼────┼──────────────────────┼───▶│  Reports            │
 │ └──────────────────┘ │    │                      │    │                     │
 │                      │    │                      │    │                     │
 └──────────────────────┘    └──────────────────────┘    └─────────────────────┘
```

Chef InSpec Profiles can be hosted from a variety of locations:

```
 ┌──────────────────────┐                                ┌─────────────────────┐
 │  Chef Infra Client   │     ┌───────────────────────┐  │    Chef Automate    │
 │                      │  ┌──│ Profiles(Supermarket, │  │                     │

>  │ ┌──────────────────┐ │  │  │ Github, local, etc)   │  │                     │

Reviewer: ```suggestion
 │ ┌──────────────────┐ │  │  │ GitHub, local, etc)   │  │                     │
```

Reviewer 2: I think I can enable Mermaid for diagramming in Hugo. Would this improve your lives? https://learn.netlify.app/en/shortcodes/mermaid/

Author: These diagrams seem to be describing "in one scenario you use a proxy/middleman" which is not really worth a lot of effort to preserve. I'll just delete them.

##### json-file

Writes the results to disk in the location specified with the [json-file location attribute](#json-file-location).

##### audit-enforcer

A special reporter that instead of sending the results of the Chef InSpec run somewhere instead will fail the Chef Infra run if any control of any Chef InSpec profile fails.

##### JSON File Location

The location on disk that Chef InSpec's json reports are saved to when using the 'json-file' reporter. Set this attribute with `default['audit']['json-file']['location']`. Defaults to: `<chef_cache_path>/compliance_reports/compliance-<timestamp>.json`

#### InSpec Backend Cache

Chef InSpec provides the ability to cache the result of commands executed on the node being tested. This drastically improves Chef InSpec performance when slower-running commands are run multiple times during execution. This feature is enabled by default for Compliance Phase. If your Chef InSpec profile runs a command multiple times expecting different output, you may have to disable this feature. To do so set the `inspec_backend_cache` attribute to `false`:

```ruby
node.normal['audit']['inspec_backend_cache'] = false
```

#### Fetchers

> Fetchers control if Chef InSpec profiles should be fetched from Chef Automate or Chef Infra Server in addition to the default fetch locations provided by Chef Inspec. Set this attribute with `default['audit']['fetcher']`. Accepted values: 'chef-server' or 'chef-automate'. With nothing specified for this attribute the [fetchers included in Chef InSpec](https://docs.chef.io/inspec/profiles/#profile-dependencies) are used. The fetchers included with Chef Infra are mutually exclusive so only one value can be specifed for this attribute.

Author: This links to a section on profile dependencies because I could not find any more appropriate documentation on how to consume pre-existing profiles in InSpec.

+++
title = "Compliance Phase"
draft = false

aliases = ["/chef_compliance_phase.html"]

[menu]
  [menu.infra]
    title = "Compliance Phase"
    identifier = "chef_infra/features/chef_compliance_phase.md Compliance Phase"
    parent = "chef_infra/features"
    weight = 80
+++

[\[edit on GitHub\]](https://github.com/chef/chef-web-docs/blob/master/content/chef_compliance_phase.md)

## Overview

> Compliance Phase enables Chef Infra to run Chef InSpec profiles as part of a client run, by downloading profiles from sources like Chef Automate, Chef Supermarket, or GitHub, and reporting the results to Chef Automate or saving the results locally. Compliance Phase has been implemented to be mostly compatible with [audit cookbook](https://github.com/chef-cookbooks/audit), so that users of the audit cookbook can remove it from the node's runlist and get the same behavior.

Reviewer: ```suggestion
Compliance Phase enables Chef Infra to run Chef InSpec profiles as part of a client run, by downloading profiles from sources like Chef Automate, Chef Supermarket, or GitHub, and reporting the results to Chef Automate or saving the results locally. Compliance Phase has been implemented to be mostly compatible with legacy [audit cookbook](https://github.com/chef-cookbooks/audit), allowing users to remove the audit cookbook from the run_list and instead set `node['audit']['compliance_phase'] = true` to enable the Compliance Phase.
```

Reviewer: We might want to just create a migration section and remove the whole sentence at the bottom here that tries to explain the migration in condensed form

## Usage

### Configuration

> Compliance Phase needs to be configured for each node where the `chef-client` runs. All node-specific configuration is done via Chef Infra attributes. Specifying one or more Chef InSpec profiles is the minimum configuration necessary to enable Compliance Phase in a Chef Infra client run. With no further configuration, this will cause a Chef Infra client run to execute the Chef InSpec profiles and write the results to disk using the default `json-file` reporter.

Reviewer: ```suggestion
Compliance Phase needs to be configured for each node where the Chef Infra Client runs. All node-specific configuration is done via Chef Infra attributes. Specifying one or more Chef InSpec profiles is the minimum configuration necessary to enable Compliance Phase in a Chef Infra Client run. With no further configuration, this will cause a Chef Infra client run to execute the Chef InSpec profiles and write the results to disk using the default `json-file` reporter.
```

> #### Profiles
> 
> Chef InSpec profiles are specified using the `node['audit']['profiles']` attribute.
> 
> ##### Automate
> 
> ```ruby
> default['audit']['profiles']['linux-baseline'] = {
>   'compliance': 'user/linux-baseline',
>   'version': '2.1.0'
> }
> ```
> 
> Note: Fetching InSpec profiles from a Chef Automate server will require additional setup for authorization and authentication.
> 
> ##### Supermarket
> 
> ```ruby
> default['audit']['profiles']['ssh'] = {
>   'supermarket': 'hardening/ssh-hardening'
> }
> ```
> 
> ##### Local paths
> 
> ```ruby
> default['audit']['profiles']['brewinc/win2012_audit'] = {
>   'path': 'E:/profiles/win2012_audit'
> }
> ```
> 
> ##### Git
> 
> ```ruby
> default['audit']['profiles']['ssl'] = {
>   'git': 'https://github.com/dev-sec/ssl-benchmark.git'
> }
> ```
> 
> ##### HTTP
> 
> ```ruby
> default['audit']['profiles']['ssh2'] = {
>   'url': 'https://github.com/dev-sec/tests-ssh-hardening/archive/master.zip'
> }
> ```
> 

Reviewer 3: This will put all this in a [tabs container](https://docs.chef.io/style_guide/#foundation-tabs-container). 

````suggestion
#### Profiles

You can fetch [Chef InSpec profiles]({{< relref "inspec/profiles/" >}}) from several different locations using the `node['audit']['profiles']` attribute.

For example:

{{< foundation_tabs tabs-id="compliance-phase-profile-panel" >}}
  {{< foundation_tab active="true" panel-link="automate-panel" tab-text="Automate">}}
  {{< foundation_tab panel-link="supermarket-panel" tab-text="Supermarket" >}}
  {{< foundation_tab panel-link="local-path-panel" tab-text="local path" >}}
  {{< foundation_tab panel-link="git-panel" tab-text="Git" >}}
  {{< foundation_tab panel-link="http-panel" tab-text="HTTP" >}}
{{< /foundation_tabs >}}

{{< foundation_tabs_panels tabs-id="compliance-phase-profile-panel" >}}
  {{< foundation_tabs_panel active="true" panel-id="automate-panel" >}}
  ```ruby
  default['audit']['profiles']['linux-baseline'] = {
    'compliance': 'user/linux-baseline',
    'version': '2.1.0'
  }
  ```

  **Note**: Fetching InSpec profiles from a Chef Automate server will require additional setup for authorization and authentication.

  {{< /foundation_tabs_panel >}}

  {{< foundation_tabs_panel panel-id="supermarket-panel" >}}
  ```ruby
  default['audit']['profiles']['ssh'] = {
    'supermarket': 'hardening/ssh-hardening'
  }
  ```

  {{< /foundation_tabs_panel >}}
  {{< foundation_tabs_panel panel-id="local-path-panel" >}}
  ```ruby
  default['audit']['profiles']['brewinc/win2012_audit'] = {
    'path': 'E:/profiles/win2012_audit'
  }
  ```
  {{< /foundation_tabs_panel >}}
  {{< foundation_tabs_panel panel-id="git-panel" >}}
  ```ruby
  default['audit']['profiles']['ssl'] = {
    'git': 'https://github.com/dev-sec/ssl-benchmark.git'
  }
  ```
  {{< /foundation_tabs_panel >}}
  {{< foundation_tabs_panel panel-id="http-panel" >}}
  ```ruby
  default['audit']['profiles']['ssh2'] = {
    'url': 'https://github.com/dev-sec/tests-ssh-hardening/archive/master.zip'
  }
  ```
  {{< /foundation_tabs_panel >}}
{{< /foundation_tabs_panels >}}

````
